include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -544,58 +544,40 @@ __pattern_minmax_element(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _
 // adjacent_find
 //------------------------------------------------------------------------
 
-template <typename _BackendTag, typename _ExecutionPolicy, typename _Iterator, typename _BinaryPredicate>
+template <typename _BackendTag, typename _ExecutionPolicy, typename _Iterator, typename _BinaryPredicate, typename _OrFirstTag>
 _Iterator
 __pattern_adjacent_find(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last,
-                        _BinaryPredicate __predicate, oneapi::dpl::__internal::__or_semantic)
+                        _BinaryPredicate __pred, _OrFirstTag __is_or_semantic)
 {
-    if (__last - __first < 2)
+    const auto __n = __last - __first;
+    if (__n < 2)
         return __last;
 
-    using _Predicate = oneapi::dpl::unseq_backend::single_match_pred<adjacent_find_fn<_BinaryPredicate>>;
+    using _Predicate = oneapi::dpl::unseq_backend::single_match_pred<_BinaryPredicate>;
+
+    auto __keep = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
+    auto __buf = __keep(__first, __last); //a scope lifetime of this instance should be.
+
+    auto __view = __buf.all_view();
+    auto __view1 = oneapi::dpl::__ranges::take_view_simple(__view, __view.size() - 1);
+    auto __view2 = oneapi::dpl::__ranges::drop_view_simple(__view, 1);
 
-    auto __keep1 = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
-    auto __buf1 = __keep1(__first, __last - 1);
-    auto __keep2 = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read, _Iterator>();
-    auto __buf2 = __keep2(__first + 1, __last);
+    using __size_calc = oneapi::dpl::__par_backend_hetero::__min_size_calc;
+    using _IndexType = std::make_unsigned_t<typename std::iterator_traits<_Iterator>::difference_type>;
+    using _TagType = std::conditional_t<__is_or_semantic(), oneapi::dpl::__par_backend_hetero::__parallel_or_tag,
+                                        oneapi::dpl::__par_backend_hetero::__parallel_find_forward_tag<_IndexType>>;
 
     // TODO: in case of conflicting names
     // __par_backend_hetero::make_wrapped_policy<__par_backend_hetero::__or_policy_wrapper>()
-    bool result = __par_backend_hetero::__parallel_find_or(
-        _BackendTag{}, ::std::forward<_ExecutionPolicy>(__exec),
-        _Predicate{adjacent_find_fn<_BinaryPredicate>{__predicate}}, __par_backend_hetero::__parallel_or_tag{},
-        oneapi::dpl::__ranges::make_zip_view(__buf1.all_view(), __buf2.all_view()));
+    auto result = __par_backend_hetero::__parallel_find_or(_BackendTag{}, std::forward<_ExecutionPolicy>(__exec),
```

Comment:
Could you apply clang-format? There are dozens of places not aligned with it. Its suggestions look good to me.